### PR TITLE
[8/x][dynamic child loading] Added bag.

### PR DESCRIPTION
**Pull request by tnowacki**
- Added bag and object bag
- added tests
- Added missing APIs to dynamic object field and object

### CODE REVIEW

**Review by oxade — APPROVED** [review]

File: crates/sui-framework/sources/bag.move [hunk 1]
```diff
@@ -0,0 +1,109 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+/// A bag is a heterogenous map-like collection. The collection is similar to `sui::table` in that
+/// it's keys and values are not stored within the `Bag` value, but instead are stored using Sui's
```
Comment (amnn) [inline crates/sui-framework/sources/bag.move:5]: nit: sp/gr.

```suggestion
/// A bag is a heterogeneous map-like collection. The collection is similar to `sui::table` in that
/// its keys and values are not stored within the `Bag` value, but instead are stored using Sui's
Commenting on lines +4 to +5
```

File: crates/sui-framework/sources/bag.move [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+/// bag::add(&mut bag1, 0, false);
+/// bag::add(&mut bag1, 1, true);
+/// bag::add(&mut bag2, 0, false);
+/// bag::add(&mut bag2, 1, true);
+/// // bag1 does not equal bag2, despite having the same entries
+/// assert!(&bag1 != &bag2, 0);
+/// ```
+/// At it's core, `sui::bag` is a wrapper around `UID` that allows for access to
+/// `sui::dynamic_field` while preventing accidentally stranding field values. A `UID` can be
+/// deleted, even if it has dynamic fields associated with it, buta bag, on the other hand, must be
```
Comment (amnn) [inline crates/sui-framework/sources/bag.move:22]: nit:
```suggestion
/// deleted, even if it has dynamic fields associated with it, but a bag, on the other hand, must be
```

File: crates/sui-framework/sources/object_bag.move [hunk 3]
```diff
@@ -0,0 +1,105 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+/// Similar to `sui::bag`, an `ObjectBag` is a heterogenous map-like collection. But unlike
+/// `sui::bag`, the values bound to these dynamic fields _must_ be objects themselves. This allows
+/// for the objects to still exist within in storage, which may be important for external tools.
```
Comment (amnn) [inline crates/sui-framework/sources/object_bag.move:6]:
```suggestion
/// for the objects to still exist in storage, which may be important for external tools.
```

File: crates/sui-framework/sources/object_bag.move [hunk 4]
```diff
@@ -0,0 +1,105 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+/// Similar to `sui::bag`, an `ObjectBag` is a heterogenous map-like collection. But unlike
```
Comment (amnn) [inline crates/sui-framework/sources/object_bag.move:4]:
```suggestion
/// Similar to `sui::bag`, an `ObjectBag` is a heterogeneous map-like collection. But unlike
```

File: crates/sui-framework/sources/object_bag.move [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+
+/// Adds a key-value pair to the bag `bag: &mut ObjectBag`
+/// Aborts with `sui::dynamic_field::EFieldAlreadyExists` if the bag already has an entry with
+/// that key `k: K`.
+public fun add<K: copy + drop + store, V: key + store>(bag: &mut ObjectBag, k: K, v: V) {
+    ofield::add(&mut bag.id, k, v);
+    bag.size = bag.size + 1;
+}
+
+/// Immutable borrows the value associated with the key in the bag `bag: &ObjectBag`.
```
Comment (amnn) [inline crates/sui-framework/sources/object_bag.move:41]:
```suggestion
/// Immutably borrows the value associated with the key in the bag `bag: &ObjectBag`.
```

**Review by amnn — COMMENTED** [review]
Sorry for the post-hoc review -- feel free to ignore the inline comments, they are more for me:  I'll fix those doc comments up when I write the docs for Bag/Table.

Main reason I wanted to comment was to ask:  Do we need both `Bag` and `ObjectBag`?  It seems, given `Bag`'s heterogeneity, that it would be okay to have a type to which you can add both wrapped and unwrapped fields.

The main pro to this idea is that it allows you to mix both, so the question is whether that's something we see people doing.  It doesn't really save any duplication, because we still need a Field and Object Field version of each API, and the main con is that it makes `Bag` not match the field and table APIs anymore.  Food for thought.

**Pull request merged**